icaruscode/PMT/OpReco/ICARUSFlashAssAna_module.cc

```diff
@@ -350,36 +537,97 @@ void opana::ICARUSFlashAssAna::analyze(art::Event const& e) {
   if( !fTriggerLabel.empty() ) { 
 
+      // Beam information
       art::Handle<std::vector<sim::BeamGateInfo>> beamgate_handle;
       e.getByLabel( fTriggerLabel, beamgate_handle );
-
+      
       if( beamgate_handle.isValid() ) {
 
         for( auto const & beamgate : *beamgate_handle ) {
 
           m_beam_gate_start = beamgate.Start(); 
           m_beam_gate_width = beamgate.Width(); 
-          m_beam_type = beamgate.BeamType() ;
+          m_beam_type = beamgate.BeamType() ; 
 
         }
 
       }
 
       else {
-        std::cout << "Invalid Trigger Data product " << fTriggerLabel.label() << "\n" ;
+        if( fDebug) { std::cout << "No sim::BeamGateInfo associated to label: " << fTriggerLabel.label() << "\n" ; }
+      }
+
+      // Now trigger information
+      art::Handle<sbn::ExtraTriggerInfo> trigger_handle;
+      e.getByLabel( fTriggerLabel, trigger_handle );
+
+      if( trigger_handle.isValid() ) {
+
+        sbn::triggerSource bit = trigger_handle->sourceType;
+
+        m_gate_type = (unsigned int)bit; 
+        m_gate_name = bitName(bit);
+        m_trigger_timestamp = trigger_handle->triggerTimestamp;
+        m_gate_start_timestamp =  trigger_handle->beamGateTimestamp;
+        m_trigger_gate_diff = m_trigger_timestamp-m_gate_start_timestamp;
+
+        std::cout << m_trigger_timestamp << "; " << m_gate_start_timestamp << " ;" << m_trigger_gate_diff << std::endl;
+
+      }
+      else{
+        if( fDebug ){ std::cout << "No raw::Trigger associated to label: " << fTriggerLabel.label() << "\n" ; }
       }
 
   }
 
   else {
-     std::cout << "Trigger Data product " << fTriggerLabel.label() << " not found!\n" ;
+     if( fDebug ){ std::cout << "Trigger Data product " << fTriggerLabel.label() << " not found!\n" ; }
   }
 
 
-  // Now we take care of the labels
+  /* 
+  Now we work on the waveforms if we are allowed to
+  */
+
+    if( !fOpDetWaveformLabels.empty() && fSaveWaveformInfo ) { 
+    
+      for( size_t i=0; i<fOpDetWaveformLabels.size(); i++ ) {
+
+        auto const label = fOpDetWaveformLabels[i];
+
+        art::Handle<std::vector<raw::OpDetWaveform>> wfm_handle;
+        e.getByLabel( label, wfm_handle );
+
+        if( wfm_handle.isValid() && !wfm_handle->empty() ) {
+
+          for( auto const & wave : *wfm_handle ){
+
+            m_channel_id = wave.ChannelNumber();
+            m_nticks = wave.Waveform().size();
+            m_baseline = Median( wave.Waveform() ); 
+            m_chargesum = std::accumulate( wave.Waveform().begin(), wave.Waveform().end(), 0, 
+             [ & ](short x, short y){ return ((m_baseline-x) + (m_baseline-y)) ; } );
+
+            fOpDetWaveformTrees[i]->Fill();
+            
+          }
+        }
+      }
+    }
+
+
+  /*
+    Now we take care of the flashes: we separate the case where we have a flash and the case where we have not a flash
+  */
+
   if ( !fFlashLabels.empty() ) {
 
     for ( size_t i=0; i<fFlashLabels.size(); i++  ) {
```

Comment:
Another comment... and another thing that you inherited: it appears that there is an implicit binding between the flash label and the cryostat it works on.
This should be checked at run time, since using the labels `[ 'E', 'W' ]` will produce different results than `[ 'W', 'E' ]` and than `[ 'E' ]` or `[ 'W' ]`.

Reply:
The binding between label and cryostat is inherited from the reconstruction of Flashes.  I think checks of some sort should be made at that level. When hits are not assigned to flashes in events with flashes the information is lost ( this is a limitation ). When flashes are not found, hits are saved on both cryostats as the ophit reconstruction is agnostic to  the cryostat where it is applied.

Reply:
I have however created a simple function to solve the only implicit binding native of this analyzer in line 713

Reply:
Yes, that was the point in question. Although rather than relying on a naming convention I would ask the data.
The protocol prescribes that only one cryostat is considered per input data product, so we can reasonably assume that all the hits in the flash collection belong to that cryostat (interestingly enough, a `recob::OpFlash` is not directly aware of which cryostat it is in). So:
```cpp
assert(!ophits.empty());
geo::CryostatID const cid = fGeom.OpDetGeoFromOpChannel(ophits.front()->OpChannel());
```
gets the ID of the cryostat the data product covers (and `cid.Cryostat` is `0` or `1` in ICARUS) from the first optical detector hit in the flash (assuming, as documented in the assertion, that a flash is associated with at least one hit). In case of error (definitely unexpected), [`geo::GeometryCore::OpDetGeoFromOpChannel()`](https://nusoft.fnal.gov/larsoft/doxsvn/html/classgeo_1_1GeometryCore.html#a4ed940e72b58f91ab8c48aafc4a3a351) throws an exception.
Here `fGeom` is an hypothetical `geo::GeometryCore const& fGeom` data member initialised in the constructor with the traditional `fGeom{ lar::providerFrom<geo::Geometry>() }` (like `geop` in some other method of the module).

Reply:
I have implemented this suggestion and removed the cryostat attribution via label